Make wxWebViewEdge compile with GCC/Clang

MaartenBent opened the pull request:
Follow-up of #2171

Improve compatibility with WRL and document how to use it.

nns52k committed:
Update webview_edge.cpp

Remove the requirement to include wrl/event.h so that GNU C++ (maybe Clang C++ as well) can compile this file without error.

nns52k committed:
Update webview_edge.h

nns52k committed:
Update webview_edge.cpp

nns52k committed:
Update webview_edge.cpp

nns52k committed:
Update webview_edge.cpp

MaartenBent committed:
Use the same function signature as the WebView2 SDK

MaartenBent committed:
Fix building with MSVC 32 bit

MaartenBent committed:
Format wxWebViewEdge Callback code

MaartenBent committed:
Use WRL when building with MSVC

vadz commented on src/msw/webview_edge.cpp:
```diff
@@ -26,10 +26,103 @@
+         * implemented to make this invokable type a concrete class instead of a
+         * abstract one.
+         */
+        wxLogWarning(_("WebView2 called unimplemented CInvokable::QueryInterface"));
```
IMO this should be a `wxLogDebug` and the message shouldn't be translated because I don't see what an end user might possibly do about this.

nns52k replied:
I am not completely sure `CInvokable::QueryInterface` will not be called by WebView2, which is why I used `wxLogWarning`. It'd better to change to `wxLogDebug` in 3.1.6 if nobody complains about the warning message when using 3.1.5 if this is going to be merged into 3.1.5.

vadz replied:
Then this should be `wxFAIL_MSG()`. Or maybe we should just implement it because it's a clear violation of COM rules to not implement QI.

nns52k replied:
Why did you said "not implement QI" when these is a QI calling `wxLogWarning`? It doesn't violate COM rules, it tries to comply with COM rules instead. Could you please do some researches yourself before popping up your so many opinions again? No offensive, but it really is annoying. I actually deleted my pull requests, because I was tired of having to explain every single trivial and OBVIOUS thing to you. Glad to see other people still interested in making (GCC/Clang + WebView2) to happen.

vadz replied:
@MaartenBent the other user is obviously wrong as a quick web search suffices to show (QI must work for `IUnknown` at the very least, that's the first point of [Raymond Chen's post](https://devblogs.microsoft.com/oldnewthing/20040326-00/?p=40033) from back in 2004 that they didn't even bother to read), but I won't discuss with them here as it's clearly counter-productive and will let you decide what to do here (of course, you can also decide to let me decide, whatever you prefer). Sorry for all the drama.

MaartenBent replied:
I put a breakpoint in `QueryInterface` and it does not seem to be called when using the webview sample. I thought it could be useful to log a message (to the developer) when this might happen in the future.

According to that post it should respond to `IID_IUnknown` and its own interface?  Maybe something like:
```c++
if (riid == __uuidof(baseT) || riid == IID_IUnknown) {
    *ppvObj = this;
    AddRef();
    return S_OK;
}
```

vadz replied:
Yes, this is what I meant by "just implement it", and I've just checked and we don't use any intermediate interfaces as bases, i.e. there are no `ICoreWebView2XXX` that inherit from anything else than `IUnknown`, so this should be correct for all of them.

Just a question: does `__uuidof()` work with gcc now? I didn't know this, but it's great news if it does.

MaartenBent replied:
> Just a question: does __uuidof() work with gcc now? 

It compiles. I suppose it will work as well. It has been part of the mingw64 header for at least 7 years.

vadz reviewed (approved):
Excellent, thanks a lot!

I'm going to merge this just changing `wxLogWarning()` to `wxLogDebug()` if you don't mind.

TcT2k commented:
Maybe the GCC specific templates should be in a separate file (something like `include/msw/private/wrl.h`) to keep the implementation file smaller and it might also of interest for other future wx code using it.

MaartenBent committed:
Implement QueryInterface of wxWebViewEdge CInvokable

MaartenBent committed:
Move Callback implementation to separate header file

TcT2k commented:
I haven't looked to much into it, but had the following idea:
Instead of the user having to modify the idl to recreate the header. Couldn't we provide a dummy _EventToken.h_ file in a include directory that would be given to the compiler during build time. As this is just required for compiling the _webview_ library and not the users application it should be under our control.

If possible this could make it easier to the user (and no additional instructions would be needed).

MaartenBent committed:
Support building wxWebViewEdge with GCC or Clang

Add EventToken.h with EventRegistrationToken implementation, that can be used by the WebView2 header.

MaartenBent commented:
Good idea, that seems to work as well (with the CMake build, I added the include dir of the build files too but haven't tested it).

PBfordev commented:
Do the AppVeyor MinGW builds use wxWebViewEdge? When trying to build this PR with MinGW I get
```
Linking CXX shared library ..\..\lib\wxmsw315ud_webview_gcc_custom.dll
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN17wxWebViewEdgeImpl20OnEnvironmentCreatedElP24ICoreWebView2Environment':
D:/Dev/Desktop/!Lib/wxWidgets-PB/src/msw/webview_edge.cpp:112: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2Environment>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI42ICoreWebView2ExecuteScriptCompletedHandlerJlPKwEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2ExecuteScriptCompletedHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI57ICoreWebView2ContainsFullScreenElementChangedEventHandlerJP13ICoreWebView2P8IUnknownEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2ContainsFullScreenElementChangedEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI39ICoreWebView2ContentLoadingEventHandlerJP13ICoreWebView2P36ICoreWebView2ContentLoadingEventArgsEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2ContentLoadingEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI45ICoreWebView2DocumentTitleChangedEventHandlerJP13ICoreWebView2P8IUnknownEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2DocumentTitleChangedEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI43ICoreWebView2NewWindowRequestedEventHandlerJP13ICoreWebView2P40ICoreWebView2NewWindowRequestedEventArgsEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2NewWindowRequestedEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI44ICoreWebView2NavigationCompletedEventHandlerJP13ICoreWebView2P41ICoreWebView2NavigationCompletedEventArgsEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2NavigationCompletedEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI38ICoreWebView2SourceChangedEventHandlerJP13ICoreWebView2P35ICoreWebView2SourceChangedEventArgsEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2SourceChangedEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI43ICoreWebView2NavigationStartingEventHandlerJP13ICoreWebView2P40ICoreWebView2NavigationStartingEventArgsEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2NavigationStartingEventHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI57ICoreWebView2CreateCoreWebView2ControllerCompletedHandlerJlP23ICoreWebView2ControllerEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2CreateCoreWebView2ControllerCompletedHandler>()'
C:/msys64/mingw32/bin/../lib/gcc/i686-w64-mingw32/10.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\wxwebview.dir/objects.a(webview_edge.cpp.obj): in function `ZN10CInvokableI58ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandlerJlP24ICoreWebView2EnvironmentEE14QueryInterfaceERK5_GUIDPPv@12':
D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/wrl/event.h:24: undefined reference to `_GUID const& __mingw_uuidof<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>()'
collect2.exe: error: ld returned 1 exit status
mingw32-make[2]: *** [libs\webview\CMakeFiles\wxwebview.dir\build.make:197: lib/wxmsw315ud_webview_gcc_custom.dll] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:1480: libs/webview/CMakeFiles/wxwebview.dir/all] Error 2
mingw32-make: *** [Makefile:148: all] Error 2
```

The compiler is from MSYS2 package _mingw-w64-i686-toolchain_ (GCC 10.2.0-6, mingw headers 9.0.0.6105.f3855e2c-1) but building with CMake-generated makefile (the only change from the default settings was turning `wxUSE_WEBVIEW_EDGE` on) from the Windows command line. I let CMake download the webview2 SDK.

BTW, despite the warning suppression macro being used, I still get
```
In file included from D:/Dev/Desktop/!Lib/wxWidgets-PB/include/wx/msw/private/webview_edge.h:18,
                 from D:\Dev\Desktop\!Lib\wxWidgets-PB\src\msw\webview_edge.cpp:27:
D:/Dev/Desktop/!Lib/wxWidgets-PB-PR2245-build-gcc10_x86_DLL/packages/Microsoft.Web.WebView2.1.0.705.50/build/native/include/WebView2.h:19: warning: ignoring '#pragma warning ' [-Wunknown-pragmas]
   19 | #pragma warning( disable: 4049 )  /* more than 64k source lines */
```

MaartenBent commented:
Hmm, you're right. I think I forgot to reenable edge after doing a clean build. I'll add the missing IDs.

I see the warning too with GCC, for Clang the suppression works. Maybe a bug in the compiler?

MaartenBent committed:
Add missing UUIDs for wxWebViewEdge

PBfordev commented:
Thanks for the quick reply and fix. I can confirm that now the CMake-based build works, and so does the webview sample. I did not test anything beyond this.

I also tried building with bundled mingw makefile (copied the webview2 sdk where needed and changed `wxUSE_WEBVIEW_EDGE` to `1` in _include/wx/msw/setup.h_) but the sample still used IE backend. I suppose makefile support has not been added yet.

Regarding the unknown pragma warning, perhaps it is related to this old-but-still-unfixed bug:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53431

MaartenBent commented:
I can add
`target_compile_options(wxwebview PRIVATE -Wno-unknown-pragmas)`
or
`set_source_files_properties("${wxSOURCE_DIR}/src/msw/webview_edge.cpp" PROPERTIES COMPILE_FLAGS -Wno-unknown-pragmas)`

I also prefer a build log without warnings.

`mingw32-make -f makefile.gcc` does build successfully with the edge backend for me, and the webview sample uses the edge backend. Did you make sure `lib/gcc_*/mswu/wx/setup.h` was updated as well? Did you copy `WebView2Loader.dll` to the webview sample build directory?

PBfordev commented:
> Did you copy `WebView2Loader.dll` to the webview sample build directory?

No, I did not. After doing that, the sample does use the edge backend. Sorry for the noise.

TcT2k commented:
@MaartenBent  
> 
> Hmm, you're right. I think I forgot to reenable edge after doing a clean build. I'll add the missing IDs.
> 
Maybe you could have a look at my changes in  #2237 and can already add the new IDs I'm using there? (shouldn't be many).

MaartenBent committed:
Silence unknown pragmas warning in the WebView2 header

Using wxGCC_WARNING_SUPPRESS(unknown-pragmas) around the webview2 header does
not work, so suppress it for the entire webview project.

MaartenBent commented:
> Maybe you could have a look at my changes in #2237 and can already add the new IDs I'm using there? (shouldn't be many).

I added `ICoreWebView2AddScriptToExecuteOnDocumentCreatedCompletedHandler` and `ICoreWebView2WebMessageReceivedEventHandler`.

For future reference, you can find these IDs before each class declaration in `MIDL_INTERFACE(...)` in `WebView2.h`. Get the hex numbers from the string and split them to the format accepted by `__CRT_UUID_DECL`.

vadz commented on include/wx/msw/wrl/EventToken.h:
```diff
@@ -0,0 +1,18 @@
+// Licence:     wxWindows licence
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef __eventtoken_h__
```
Should we use `_WX_MSW_WRL_EVENTTOKEN_H_` here or it important to define this header guard?

MaartenBent replied:
I used the same guard name as in the Microsoft wrl header, so there will be no type redefinition if for some reason both headers are included.

vadz replied:
Yes, this is what I suspected, but wasn't sure about it. Maybe it could be worth adding a brief comment explaining it here just to prevent somebody (yes, I often speak of myself in third person, why?) from normalizing it later without thinking?

MaartenBent replied:
Sure, I'll add a comment

vadz commented on build/bakefiles/multilib.bkl:
```diff
@@ -180,6 +180,14 @@
+    </set>
+
+    <set var="webview_edge_pragma_warning">
+        <if cond="TOOLKIT=='MSW' and IS_MSVC=='0'">-Wno-unknown-pragmas</if>
```
I don't have any problems with doing it like this, but using `pragma GCC diagnostic push/ignored/pop` before/after the code containing the unknown pragmas could have been an even simpler solution.

MaartenBent replied:
That is what I did initially, see https://github.com/wxWidgets/wxWidgets/commit/e67b0af7601a3d49b61636034cc6d9486a01ea9f#diff-3c03e1f5645f42320bb127a440d48aa10b9c1f70179df0a3f316fe022abc71f6

But this did not work for GCC, the warning still shows.

vadz replied:
Sorry for missing this, I now see PB has even found the corresponding gcc bug already, but I didn't realize its import.

vadz reviewed (commented):
Thanks again for your work on this!

Should this be merged already (perhaps with a tiny rename of the header guard) or do you have any more planned changes here?

MaartenBent commented:
I'll add one more commit with documentation update in the next minutes, then it can be merged.

MaartenBent committed:
Mention GCC/Clang support in wxWebViewEdge documentation

And add a comment explaining the name of the EventRegistrationToken include guard.

vadz commented:
Thanks, merged now after some squashing/rewording of the commits.
